### PR TITLE
Make `ray` an optional depencency

### PR DESCRIPTION
**Pull request by flo-ri-an**
This PR intends to make `ray` an optional dependency. Due to the large size of ray and all its dependencies, this helps to bring down container size when `ray` is not needed.

hiclass will fallback to joblib.Parallel if ray is not installed and `n_jobs` is not 1.

I wrote a couple of test cases which pass, along with all tests that passed before the changes.

This PR at the current state is most likely not ready to be merged. Treat it as a suggestion for the time being.

Main areas of uncertainty:
- I did not find a lot of documentation on how to specify extra packages in a Pipfile.
- I have not tested in an actual use-case as of yet.

### CODE REVIEW

**Comment by mirand863** [comment]
Hi @flo-ri-an,

Thank you for  the contribution! I will take a look into it. Out of curiosity, what is the difference in containers size with and without Ray?

File: tests/test_LocalClassifierPerLevel.py [hunk 1]
```diff
@@ -58,6 +58,25 @@ def test_fit_digraph(digraph_logistic_regression):
             pytest.fail(repr(e))
     assert 1
 
+def test_fit_digraph_joblib_parallel(digraph_logistic_regression):
```
Comment (mirand863) [inline tests/test_LocalClassifierPerLevel.py:61]:
```suggestion
def test_fit_digraph_joblib_multiprocessing(digraph_logistic_regression):
```

File: tests/test_LocalClassifierPerNode.py [hunk 2]
```diff
@@ -110,6 +110,28 @@ def test_fit_digraph(digraph_logistic_regression):
             pytest.fail(repr(e))
     assert 1
 
+def test_fit_digraph_joblib_multiplrocessing(digraph_logistic_regression):
```
Comment (mirand863) [inline tests/test_LocalClassifierPerNode.py:113]:
```suggestion
def test_fit_digraph_joblib_multiprocessing(digraph_logistic_regression):
```

**Review by mirand863 — CHANGES_REQUESTED** [review]
It looks great and ray still works on a larger test I did with the local classifier per node. Can you please apply these 2 minor suggestions and run `black hiclass tests` to make sure it will not crash the pipeline? I will see if I can fix the pipeline to work on pull requests from other people.

PS: I did have a problem installing it with `pip install -e .`:
```
Obtaining file:///home/fabio/Desktop/optional_ray_hiclass
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      error in hiclass setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

File: setup.py [hunk 3]
```diff
@@ -26,11 +26,11 @@
 KEYWORDS = ["hierarchical classification"]
 DACS_SOFTWARE = "https://gitlab.com/dacs-hpi"
 # What packages are required for this module to be executed?
-REQUIRED = ["networkx", "numpy", "scikit-learn", "ray", "protobuf<4.0"]
+REQUIRED = ["networkx", "numpy", "scikit-learn", "protobuf<4.0"]
 
 # What packages are optional?
 # 'fancy feature': ['django'],}
-EXTRAS = {}
+EXTRAS = {"ray"}
```
Comment (mirand863) [inline setup.py:33]:
```suggestion
EXTRAS = {"ray": ["ray>=1.11.0"]}
```

**Review by mirand863 — CHANGES_REQUESTED** [review]
This fixes the issue when installing with `pip install -e .`

**Comment by mirand863** [comment]
> I don't know exactly, MLOps just told me that it's quite hefty. I think ~150MB for ray alone, plus possibly some more dependencies? Also, a big issue seems to be that the pypi-package does not distribute sources, which makes it difficult to optimize/cache/build from source. A related issue is here: [ray-project/ray#6246](https://github.com/ray-project/ray/issues/6246)
> 
> Ray will eventually be the way forward on our end as well. But until we're ready, it would be nice to get by with joblib.

Perhaps using the conda installation could work for you. I know they distribute some files, e.g., [https://anaconda.org/conda-forge/hiclass/files](https://anaconda.org/conda-forge/hiclass/files), but I am not 100% sure if you can use those to cache.

**Comment by flo-ri-an** [comment]
> > I don't know exactly, MLOps just told me that it's quite hefty. I think ~150MB for ray alone, plus possibly some more dependencies? Also, a big issue seems to be that the pypi-package does not distribute sources, which makes it difficult to optimize/cache/build from source. A related issue is here: [ray-project/ray#6246](https://github.com/ray-project/ray/issues/6246)
> > Ray will eventually be the way forward on our end as well. But until we're ready, it would be nice to get by with joblib.
> 
> Perhaps using the conda installation could work for you. I know they distribute some files, e.g., https://anaconda.org/conda-forge/hiclass/files, but I am not 100% sure if you can use those to cache.
> 
> In my benchmark it took the same amount of time with joblib and it saved 1GB of RAM when compared with Ray. Good job!

Awesome!

We use `poetry` everywhere for dependency management, so it would be quite some effort to switch over to conda. Also, imho poetry is the most advanced dependency manager currently out there.
We'll have to find some way around this issue once we start using ray.

**Comment by flo-ri-an** [comment]
> 

`black hiclass tests` succeeds for me locally. 

```
All done! ✨ 🍰 ✨
17 files left unchanged.
```

**Comment by mirand863** [comment]
> > 
> 
> `black hiclass tests` succeeds for me locally.
> 
> ```
> All done! ✨ 🍰 ✨
> 17 files left unchanged.
> ```

@flo-ri-an Perhaps we have different versions of black. There were some lines that had to be added or removed. I also ran some more installation tests and everything worked fine, so I will merge this PR. Thank you again!

**Comment by flo-ri-an** [comment]
Great, thank you as well for the speedy resolution.

**Pull request merged**